File: take_u_forward/graph/21_Course Schedule I.java
class Solution {
    public boolean canFinish(int n, int[][] prerequisites) {
        // code here
        // Step 1: Create adjacency list
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }

        // Step 2: Build graph (y -> x)
        for (int[] p : prerequisites) {
            int x = p[0];
            int y = p[1];
            adj.get(y).add(x);
        }

        // Step 3: Compute indegree
        int[] indegree = new int[n];
        for (int i = 0; i < n; i++) {
            for (int it : adj.get(i)) {
                indegree[it]++;
            }
        }

        // Step 4: Queue for courses with indegree 0
        Queue<Integer> q = new ArrayDeque<>();
        for (int i = 0; i < n; i++) {
            if (indegree[i] == 0) {
                q.add(i);
            }
        }

        // Step 5: Kahn's Algorithm
        int count = 0;
        while (!q.isEmpty()) {
            int node = q.poll();
            count++;

            for (int it : adj.get(node)) {
                indegree[it]--;
                if (indegree[it] == 0) {
                    q.add(it);
                }
            }
        }

        // Step 6: If all courses processed → possible
        return count == n;
    }
}

File: take_u_forward/graph/21_Course Schedule I.cpp
class Solution {
  public:
    bool canFinish(int n, vector<vector<int>>& prerequisites) {
        // Code here
         // Step 1: Create adjacency list
        vector<vector<int>> adj(n);
        for (auto &p : prerequisites) {
            int x = p[0];
            int y = p[1];
            adj[y].push_back(x);   // y -> x
        }

        // Step 2: Compute indegree
        vector<int> indegree(n, 0);
        for (int i = 0; i < n; i++) {
            for (int it : adj[i]) {
                indegree[it]++;
            }
        }

        // Step 3: Push nodes with indegree 0
        queue<int> q;
        for (int i = 0; i < n; i++) {
            if (indegree[i] == 0) {
                q.push(i);
            }
        }

        // Step 4: Kahn's Algorithm
        int count = 0;
        while (!q.empty()) {
            int node = q.front();
            q.pop();
            count++;

            for (int it : adj[node]) {
                indegree[it]--;
                if (indegree[it] == 0) {
                    q.push(it);
                }
            }
        }

        // Step 5: Check if all courses are processed
        return count == n;
    }
};